### DISCUSSION TITLE
What is crossgen2?

### DISCUSSION THREAD

**Discussion by minxuanz** [dotnet/runtime#95171] [Q&A] [answered]
![183845](https://github.com/dotnet/runtime/assets/45393763/dee55b1d-51e4-4de5-bd91-0bd2e1399ffe)

I build runtime  on linux first.I find a file named crossgen2.

google say 
>Crossgen2 is a tool that allows you to generate unique, native code in a ready-to-run format

It means crossgen2 can generate native code without jit?
 `gcc -o a.out a.c` like a.out?

> /home/magus/runtime/dotnet.sh /home/magus/runtime/artifacts/bin/coreclr/linux.x64.Debug/crossgen2/crossgen2.dll -o:/home/magus/runtime/artifacts/bin/coreclr/linux.x64.Debug/System.Private.CoreLib.dll -r:/home/magus/runtime/artifacts/bin/coreclr/linux.x64.Debug/IL/*.dll --targetarch:x64 --targetos:linux -O --verify-type-and-field-layout /home/magus/runtime/artifacts/bin/coreclr/linux.x64.Debug/IL/System.Private.CoreLib.dll --perfmap-format-version:1 --perfmap --perfmap-path:/home/magus/runtime/artifacts/bin/coreclr/linux.x64.Debug

I notice System.Private.CoreLib.dll  use crossgen2 , it generates System.Private.CoreLib.ni.r2rmap and  System.Private.CoreLib.dll.
but  /home/magus/runtime/artifacts/bin/coreclr/linux.x64.Debug/System.Private.CoreLib.dll   is still IL code ,not native code.
Sorry my english is not good

**Comment by hez2010** [comment, marked as answer]
It's an ahead-of-time compiler which generates native code and saves it into the managed assembly in a R2R format, so that the runtime can load and execute those native code from the assembly directly on startup without need of jitting them. Basically, it is used for speeding up the startup. 

Note that crossgen2 won't generate native code for all your code (as some are versionable while some others require dynamic code generation or runtime generic instantiation, etc.), so it still needs IL code for JIT.

> is still IL code ,not native code

It's a mix of IL code and native code

If you are looking into a way to build your app and run it without the need of JIT, you can check out ilc, which is the compiler used by NativeAOT. While in fact crossgen2 and ilc are sharing a lot of code.
